http-clients/src/main/java/com/palantir/remoting3/clients/ClientConfiguration.java

```diff
@@ -61,6 +64,15 @@
     /** The credentials to use for the proxy selected by {@link #proxy}. */
     Optional<BasicCredentials> proxyCredentials();
 
+    /**
+     * Rewrites all request URLs to the mesh proxy, but keeps the "Host" (or HTTP2 :authority pseudo-header) as the
+     * authority from the original request URL.
+     * <p>
+     * Note that if this option is set, then the {@link #maxNumRetries} must also be set to 0 since the mesh proxy is
+     * expected to handle all retry logic.
```

Comment:
The proxy retries even for streamed requests?

Reply:
Ah no it doesn't - but we really shouldn't be retrying for streamed requested in http-remoting either most likely and should let the application retry. This is since the Request object in OkHttp is an InputStream.

Reply:
Request's body*

Reply:
Ah actually MultipartBody has a `writeTo(BufferedSink sink)` method.

Reply:
It's questionable if we should transparently retry streams imo - it could lead to some pretty questionable performance problems.

Reply:
How do retries of very large requests work? Say a request that’s 2MB?

Reply:
I am not too sure how this works with HTTP 1.1 services, but in HTTP 2 the upstream can actually respond before the client has written the body of the request. Since envoy does HTTP 2 streams between all the envoy nodes I imagine they'd implement it so that the rate limiter responds immediately before the request body arrives.

Reply:
Will have to do a bit of investigation on our test stack. [Here's the envoy documentation for retries
](https://www.envoyproxy.io/envoy/configuration/http_filters/router_filter.html#config-http-filters-router-x-envoy-retry-on) - it's reasonably sparse about the details.

Reply:
I think it’s optimistic to outlaw retries usage here, I guess.

Reply:
The issue if you don't disable retries is then you get n * m retries. Perhaps we want to not require that people specify a number here so that we can tweak it later.

Another change we can do at a later date is to support annotating methods as idempotent, and only retrying on idempotent methods. Or, perhaps, assuming idempotent as default.

Reply:
Sure, forget it.